File: src/data/tutorials/065-regular-expression-matching.js
/**
 * Tutorial content for #065 — Regular Expression Matching
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Implement regular expression matching where '.' matches any single character and '*' matches zero or more of the preceding element. The match must cover the entire string s.`,
    example: `s = "aa", p = "a*"\n→ "a*" means "zero or more a's" → matches "aa"\n✅ Answer: true`,
    keyInsight: `The '*' is the tricky part. When you see '*', look at the character before it (call it c). You can either skip the "c*" pattern entirely (zero uses) or match one c and keep the pattern (one or more uses). Handle '.' as matching anything.`,
  },

  approaches: {
    'Recursive (no memoization)': {
      intuition: `Pure recursion to show the structure before optimisation. At each step, if p[1]=='*', try using zero c's (skip c*) or one c followed by more (if s[0] matches c). Otherwise, match single characters and recurse.`,
      steps: [
        `Base case: if pattern is empty, return s is empty.`,
        `First char matches = (p is non-empty) and (p[0]=='.' or p[0]==s[0]).`,
        `If p.length >= 2 and p[1] == '*':`,
        `  return isMatch(s, p.substring(2))  ← zero uses of c*`,
        `       OR (firstCharMatches AND isMatch(s.substring(1), p)).  ← one use`,
        `Else: return firstCharMatches AND isMatch(s.substring(1), p.substring(1)).`,
      ],
      example: `s="aa", p="a*"\n\nisMatch("aa","a*"):\n  p[1]='*' → try zero: isMatch("aa","") → false\n           try one:  'a'=='a' AND isMatch("a","a*")\nisMatch("a","a*"):\n  try zero: isMatch("a","") → false\n  try one:  'a'=='a' AND isMatch("","a*")\nisMatch("","a*"):\n  try zero: isMatch("","") → true ✅\n✅ Answer: true`,
      keyInsight: `O(2^(s+p)) worst case. Never use without memoisation — shown only to understand the recurrence.`,
    },

    'Memoization (Top-Down)': {
      intuition: `Same recursive logic, but cache each (i, j) pair — i = position in s, j = position in p. Many sub-problems repeat (especially with patterns like "a*a*a*"), so memoisation cuts exponential to polynomial time.`,
      steps: [
        `Create memo[m+1][n+1] filled with -1.`,
        `Base case: if j == n, return i == m.`,
        `First-char-matches: i < m and (p[j]=='.' or p[j]==s[i]).`,
        `If j+1 < n and p[j+1]=='*':`,
        `  memo[i][j] = solve(i,j+2) OR (firstMatch AND solve(i+1,j)).`,
        `Else: memo[i][j] = firstMatch AND solve(i+1,j+1).`,
        `Return memo[i][j].`,
      ],
      example: `s="aa", p="a*"\n\nsolve(0,0): p[1]='*' → solve(0,2) OR ('a'=='a' AND solve(1,0))\nsolve(0,2): j=2==n=2, i=0≠m=2 → false\nsolve(1,0): p[1]='*' → solve(1,2) OR ('a'=='a' AND solve(2,0))\nsolve(1,2): false\nsolve(2,0): j=2==n, i=2==m → true ✅\n✅ Answer: true`,
      keyInsight: `O(m×n) time and space. Each (i,j) pair is computed exactly once — the standard solution for this problem.`,
    },

    '2D Bottom-Up DP (string-prefix prefix)': {
      intuition: `Build a 2D table dp[i][j] = "does s[0..i-1] match p[0..j-1]?". Fill from the base case dp[0][0]=true outward. The '*' case sets dp[i][j] based on either zero uses (dp[i][j-2]) or one+ uses (dp[i-1][j] if chars match).`,
      steps: [
        `Create dp[m+1][n+1]; dp[0][0] = true.`,
        `Fill dp[0][j]: patterns like "a*", "a*b*" can match empty s.`,
        `  If p[j-1]=='*': dp[0][j] = dp[0][j-2].`,
        `For i from 1 to m, j from 1 to n:`,
        `  If p[j-1]=='*': dp[i][j] = dp[i][j-2]  (zero uses)`,
        `    OR (p[j-2]=='.' or p[j-2]==s[i-1]) AND dp[i-1][j].  (one+ uses)`,
        `  Else if p[j-1]=='.' or p[j-1]==s[i-1]: dp[i][j] = dp[i-1][j-1].`,
        `Return dp[m][n].`,
      ],
      example: `s="aa", p="a*"\n\n    ""  a  *\n""   T  F  T\na    F  T  T\na    F  F  T\n\ndp[2][2] = true\n✅ Answer: true`,
      keyInsight: `O(m×n) time and space. Bottom-up avoids recursion overhead and is easy to trace with the table.`,
    },

    'Space-Optimized DP (1-D rolling array)': {
      intuition: `dp[i][j] only needs the previous row (dp[i-1][...]) and the current row (dp[i][...]). Replace the 2D table with two 1D arrays. Process row by row, swapping after each.`,
      steps: [
        `Create prev[n+1] and curr[n+1]; initialise prev for the empty-string row.`,
        `For i from 1 to m:`,
        `  curr[0] = false.`,
        `  For j from 1 to n:`,
        `    Fill curr[j] using same logic as 2D DP, reading prev for dp[i-1][...] values.`,
        `  Swap prev and curr.`,
        `Return prev[n].`,
      ],
      example: `s="aa", p="a*"\n\nprev=[T,F,T]  (empty string vs patterns)\nRow 'a' (i=1):\n  j=1 ('a'): 'a'=='a' → curr[1]=prev[0]=T\n  j=2 ('*'): zero uses → curr[2]=curr[0]=F; one use: 'a'=='a' AND prev[2]=T → curr[2]=T\nprev=[F,T,T]\nRow 'a' (i=2): same logic → prev=[F,F,T]\nReturn prev[2]=true\n✅ Answer: true`,
      keyInsight: `O(m×n) time, O(n) space. Use this when the string lengths are large and 2D memory is a concern.`,
    },
  },
}

File: public/cpp/065-regular-expression-matching.cpp
/*
 * Problem: Regular Expression Matching (LeetCode 10)
 * Link: https://leetcode.com/problems/regular-expression-matching/
 * Difficulty: Hard
 * Category: Dynamic Programming
 *
 * Description:
 * Given an input string s and a pattern p, implement regular expression matching
 * with support for '.' and '*' where:
 * - '.' matches any single character.
 * - '*' matches zero or more of the preceding element.
 * The matching should cover the entire input string (not partial).
 *
 * Example 1: s="aa", p="a*"  → true  ('a*' matches "aa")
 * Example 2: s="ab", p=".*"  → true  ('.*' matches "ab")
 * Example 3: s="aa", p="a"   → false
 * Example 4: s="aab",p="c*a*b" → true
 *
 * Constraints:
 * - 1 <= s.length <= 20
 * - 1 <= p.length <= 30
 * - s contains only lowercase letters
 * - p contains only lowercase letters, '.', and '*'
 * - '*' never appears without a valid preceding element
 */

#include <vector>
#include <string>
#include <iostream>
using namespace std;

/*
 * APPROACH 1: 2D DP (Bottom-Up)
 *
 * dp[i][j] = does s[0..i-1] match p[0..j-1]?
 *
 * Recurrence:
 *   if p[j-1] == s[i-1] or p[j-1] == '.':
 *     dp[i][j] = dp[i-1][j-1]
 *   elif p[j-1] == '*':
 *     dp[i][j] = dp[i][j-2]              // use '*' as zero occurrences
 *             || (p[j-2] matches s[i-1]   // use '*' as one-more occurrence
 *                && dp[i-1][j])
 *
 * Time:  O(m * n)  where m=|s|, n=|p|
 * Space: O(m * n)
 */
class Solution1 {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<vector<bool>> dp(m+1, vector<bool>(n+1, false));
        dp[0][0] = true;
        // Handle patterns like a*, a*b*, a*b*c* that can match empty string
        for (int j = 2; j <= n; j++)
            if (p[j-1] == '*') dp[0][j] = dp[0][j-2];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (p[j-1] == '*') {
                    dp[i][j] = dp[i][j-2]; // zero occurrences
                    if (p[j-2] == s[i-1] || p[j-2] == '.')
                        dp[i][j] = dp[i][j] || dp[i-1][j]; // one more
                } else if (p[j-1] == s[i-1] || p[j-1] == '.') {
                    dp[i][j] = dp[i-1][j-1];
                }
            }
        }
        return dp[m][n];
    }
};

/*
 * APPROACH 2: Space-Optimised DP (1D rolling array)
 *
 * Only need previous row → use two 1D arrays.
 *
 * Time:  O(m * n)
 * Space: O(n)
 */
class Solution2 {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<bool> prev(n+1, false), curr(n+1, false);
        prev[0] = true;
        for (int j = 2; j <= n; j++)
            if (p[j-1] == '*') prev[j] = prev[j-2];
        for (int i = 1; i <= m; i++) {
            fill(curr.begin(), curr.end(), false);
            curr[0] = false;
            for (int j = 1; j <= n; j++) {
                if (p[j-1] == '*') {
                    curr[j] = (j >= 2) ? curr[j-2] : false;
                    if (p[j-2] == s[i-1] || p[j-2] == '.')
                        curr[j] = curr[j] || prev[j];
                } else if (p[j-1] == s[i-1] || p[j-1] == '.') {
                    curr[j] = prev[j-1];
                }
            }
            swap(prev, curr);
        }
        return prev[n];
    }
};

/*
 * APPROACH 3: Memoization (Top-Down)
 *
 * Recursively match s[i..] against p[j..], cache results.
 *
 * Time:  O(m * n)
 * Space: O(m * n) + stack
 */
class Solution3 {
    vector<vector<int>> memo;
    bool helper(const string& s, const string& p, int i, int j) {
        if (j == (int)p.size()) return i == (int)s.size();
        if (memo[i][j] != -1) return memo[i][j];
        bool firstMatch = (i < (int)s.size() && (p[j] == s[i] || p[j] == '.'));
        bool res;
        if (j+1 < (int)p.size() && p[j+1] == '*') {
            res = helper(s, p, i, j+2) ||           // skip x*
                  (firstMatch && helper(s, p, i+1, j)); // consume one
        } else {
            res = firstMatch && helper(s, p, i+1, j+1);
        }
        return memo[i][j] = res;
    }
public:
    bool isMatch(string s, string p) {
        memo.assign(s.size()+1, vector<int>(p.size()+1, -1));
        return helper(s, p, 0, 0);
    }
};

/*
 * APPROACH 4: Pure Recursion (for understanding)
 *
 * Same as Approach 3 but without caching.
 * Exponential time — fine only for short strings.
 *
 * Time:  O(2^(m+n)) worst case
 * Space: O(m+n) recursion stack
 */
class Solution4 {
    bool helper(const string& s, const string& p, int i, int j) {
        if (j == (int)p.size()) return i == (int)s.size();
        bool firstMatch = (i < (int)s.size() && (p[j] == s[i] || p[j] == '.'));
        if (j+1 < (int)p.size() && p[j+1] == '*') {
            return helper(s, p, i, j+2) ||
                   (firstMatch && helper(s, p, i+1, j));
        }
        return firstMatch && helper(s, p, i+1, j+1);
    }
public:
    bool isMatch(string s, string p) { return helper(s, p, 0, 0); }
};

void test(string s, string p, int approach) {
    bool result;
    cout << "s=\"" << s << "\" p=\"" << p << "\" ";
    switch(approach) {
        case 1: { Solution1 sol; result = sol.isMatch(s,p); cout<<"2D-DP:  "; break; }
        case 2: { Solution2 sol; result = sol.isMatch(s,p); cout<<"1D-DP:  "; break; }
        case 3: { Solution3 sol; result = sol.isMatch(s,p); cout<<"Memo:   "; break; }
        case 4: { Solution4 sol; result = sol.isMatch(s,p); cout<<"Recur:  "; break; }
    }
    cout << (result ? "true" : "false") << "\n";
}

int main() {
    for(int i=1;i<=4;i++) test("aa","a",i);
    for(int i=1;i<=4;i++) test("aa","a*",i);
    for(int i=1;i<=4;i++) test("ab",".*",i);
    for(int i=1;i<=4;i++) test("aab","c*a*b",i);
    for(int i=1;i<=4;i++) test("mississippi","mis*is*p*.",i);
    return 0;
}

/*
 * COMPLEXITY SUMMARY:
 * Approach 1 (2D DP):    Time O(m*n)    Space O(m*n)
 * Approach 2 (1D DP):    Time O(m*n)    Space O(n)
 * Approach 3 (Memo):     Time O(m*n)    Space O(m*n)+stack
 * Approach 4 (Recursion):Time O(2^(m+n)) Space O(m+n)
 */

// Made with Bob
